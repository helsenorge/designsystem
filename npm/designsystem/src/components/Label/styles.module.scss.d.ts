export type Styles = {
  'after-label-children': string;
  label: string;
  'label--on-dark': string;
  'label--semibold': string;
  'label--sublabel': string;
  'label-content-wrapper': string;
  'label-wrapper': string;
  'label-wrapper--after-label-children': string;
  'label-wrapper--no-bottom-margin': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
