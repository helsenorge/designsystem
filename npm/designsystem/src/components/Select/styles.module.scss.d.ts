export type Styles = {
  select: string;
  'select--invalid': string;
  'select--on-blueberry': string;
  'select-arrow': string;
  'select-grid': string;
  'select-grid--disabled': string;
  'select-grid--invalid': string;
  'select-grid--on-blueberry': string;
  'select-grid--transparent': string;
  'select-wrapper': string;
  'select-wrapper__after-label-children': string;
  'select-wrapper__label-wrapper': string;
  'select-wrapper__label-wrapper--on-dark': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
