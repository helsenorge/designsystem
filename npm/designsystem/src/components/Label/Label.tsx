import React, { FunctionComponent } from 'react';

import cn from 'classnames';

import { Sublabel, SublabelProps } from './SubLabel';
import { AnalyticsId, FormOnColor } from '../../constants';
import { isComponent } from '../../utils/component';
import FormFieldTag, { FormFieldTagLevel } from '../FormFieldTag';
import Spacer from '../Spacer';
import StatusDot, { StatusDotProps } from '../StatusDot';

import styles from './styles.module.scss';

export type LabelText = {
  hideFromScreenReader?: boolean;
  text: string;
  type?: 'normal' | 'subdued';
};

export type LabelTags = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span' | 'label' | 'p';

export interface LabelProps {
  /** Component shown after label - discourage use of this */
  afterLabelChildren?: React.ReactNode;
  /** Sets the content of the Label */
  children?: React.ReactNode;
  /** Sets a tag that describes whether the form field is required or optional */
  formFieldLevel?: FormFieldTagLevel;
  /** Adds custom classes to the label tag. */
  labelClassName?: string;
  /** Adds custom classes to the label text. */
  labelTextClassName?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Id that is put on the "for" attribute of the label */
  htmlFor?: string;
  /** Changes the underlying element of the label */
  htmlMarkup?: LabelTags;
  /** Id som plasseres p√• <label/> */
  labelId?: string;
  /** Array of main label strings. Can be of type semibold or normal */
  labelTexts?: LabelText[];
  /** Array of sublabel strings. Can be of type semibold or normal */
  onColor?: keyof typeof FormOnColor;
  /** StatusDot placed underneath the last sublabel */
  statusDot?: React.ReactNode;
  /** Sublabel component */
  sublabel?: React.ReactNode;
  /** Adds custom classes to the div wrapping the sublabels. */
  sublabelWrapperClassName?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
}

export const getLabelText = (label: React.ReactNode): string => {
  let allLabelText = '';

  if (isComponent<LabelProps>(label, Label)) {
    label.props.labelTexts?.forEach(labelText => {
      allLabelText += !labelText.hideFromScreenReader ? labelText.text : '';
    });
  }

  return allLabelText;
};

export const renderLabel = (label: React.ReactNode, inputId: string, onColor: FormOnColor, markup?: LabelTags): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            htmlMarkup: markup || 'label',
            onColor,
          })
        : typeof label === 'string' && <Label labelTexts={[{ text: label, type: 'normal' }]} htmlFor={inputId} onColor={onColor} />}
    </>
  );
};

export const renderLabelAsParent = (
  label: React.ReactNode,
  children: React.ReactNode,
  inputId: string,
  onColor: FormOnColor,
  labelClassName?: string,
  labelTextClassName?: string,
  sublabelWrapperClassName?: string,
  large?: boolean,
  markup?: LabelTags
): React.ReactNode => {
  return (
    <>
      {label && isComponent<LabelProps>(label, Label)
        ? React.cloneElement(label, {
            htmlFor: inputId,
            onColor,
            children: children,
            labelClassName: cn(labelClassName, label.props.labelClassName),
            labelTextClassName: labelTextClassName,
            htmlMarkup: markup || 'label',
            sublabelWrapperClassName: sublabelWrapperClassName,
            sublabel: large ? undefined : label.props.sublabel,
            statusDot: large ? undefined : label.props.statusDot,
          })
        : typeof label === 'string' && (
            <Label
              labelTexts={[{ text: label, type: 'subdued' }]}
              htmlFor={inputId}
              onColor={onColor}
              htmlMarkup={markup || 'label'}
              labelClassName={labelClassName}
              labelTextClassName={labelTextClassName}
              sublabelWrapperClassName={sublabelWrapperClassName}
            >
              {children}
            </Label>
          )}
    </>
  );
};

const Label: FunctionComponent<LabelProps> = ({
  afterLabelChildren,
  children,
  className,
  formFieldLevel,
  htmlFor,
  htmlMarkup = 'label',
  labelClassName,
  labelTextClassName,
  labelId,
  labelTexts,
  onColor = FormOnColor.onwhite,
  statusDot,
  sublabel,
  sublabelWrapperClassName,
  testId,
}) => {
  const hasChildren = children && typeof children !== 'undefined';
  const labelWrapperClasses = cn(
    styles['label-wrapper'],
    { [styles['label-wrapper--no-bottom-margin']]: hasChildren, [styles['label-wrapper--after-label-children']]: afterLabelChildren },
    className
  );

  const mapLabels = (): React.ReactNode => {
    if (typeof labelTexts === 'undefined') return null;

    return labelTexts.map((labelText, index) => {
      const labelClasses = cn(
        styles.label,
        {
          [styles['label--subdued']]: labelText.type === 'subdued',
          [styles['label--on-dark']]: onColor === FormOnColor.ondark,
        },
        labelTextClassName
      );
      return (
        <span aria-hidden={labelText.hideFromScreenReader} className={labelClasses} key={index}>
          {labelText.text}
        </span>
      );
    });
  };
  const CustomTag = htmlMarkup;

  return (
    <div className={labelWrapperClasses}>
      <div>
        <CustomTag className={labelClassName} id={labelId} htmlFor={htmlFor} data-testid={testId} data-analyticsid={AnalyticsId.Label}>
          <span className={styles['label-content-wrapper']}>
            {children}
            <span className={styles.label__texts}>{mapLabels()}</span>
          </span>
        </CustomTag>
        <div className={sublabelWrapperClassName}>
          {formFieldLevel && <FormFieldTag level={formFieldLevel} />}
          {sublabel &&
            isComponent<SublabelProps>(sublabel, Sublabel) &&
            React.cloneElement(sublabel, {
              onColor: onColor as FormOnColor,
            })}
          {statusDot && isComponent<StatusDotProps>(statusDot, StatusDot) && (
            <>
              <Spacer size={'3xs'} />
              {React.cloneElement(statusDot, {
                onColor: onColor === FormOnColor.ondark ? 'ondark' : 'onwhite',
              })}
            </>
          )}
        </div>
      </div>
      {afterLabelChildren && <div className={styles['after-label-children']}>{afterLabelChildren}</div>}
    </div>
  );
};

export default Label;
