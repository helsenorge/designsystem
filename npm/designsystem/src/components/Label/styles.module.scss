@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';
@import '../../scss/font-settings';

.label-wrapper {
  margin-bottom: getSpacer(2xs);
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    line-height: $lineheight-size-sm;
  }

  &--no-bottom-margin {
    margin-bottom: 0;
  }

  &--after-label-children {
    display: flex;
    align-items: flex-end;
  }
}

.label-content-wrapper {
  display: flex;
}

.label {
  font-size: $font-size-sm;
  font-weight: 400;
  line-height: $lineheight-size-xs;

  @media (min-width: map.get($grid-breakpoints, sm)) {
    font-size: $font-size-md;
    line-height: $lineheight-size-sm;
  }

  &--semibold {
    font-weight: 600;
  }

  &--sublabel {
    display: block;
    font-size: $font-size-xs;
    line-height: 1.25rem;
    color: $neutral800;

    @media (min-width: map.get($grid-breakpoints, sm)) {
      font-size: $font-size-sm;
      line-height: $lineheight-size-xs;
    }
  }

  &--on-dark {
    color: $white;
  }

  &::after {
    content: ' ';
  }
}

.after-label-children {
  padding-left: getSpacer(xs);
}
