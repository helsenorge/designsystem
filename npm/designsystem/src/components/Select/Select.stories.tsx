import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import { palette } from '../../theme/palette';
import Select from './Select';

export default {
  title: 'Components/Select',
  component: Select,
  parameters: {
    docs: {
      description: {
        component:
          'Som innbygger vil jeg kunne se hva som er valgt av et liste med skjulte valgalternativer, og åpne den skjulte listen og gjøre andre valg dersom jeg ønsker det, slik at jeg kan løse mine oppgaver. <br><br>Nytteverdi: Lar innbygger velge ETT blant flere alternativer <br>Bruksområde: Brukes når innbygger skal kunne velge ett enkelt valg under en etikett, men plassen som er til rådighet for kontrollen bare rommer ett valg. ',
      },
    },
  },
  argTypes: {
    disabled: {
      control: 'boolean',
      defaultValue: false,
    },
  },
} as ComponentMeta<typeof Select>;

export const Default: ComponentStory<typeof Select> = (args: any) => (
  <>
    <Select {...args} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
  </>
);

export const Concepts: ComponentStory<typeof Select> = (args: any) => (
  <div style={{ backgroundColor: palette.blueberry100, padding: '2rem' }}>
    <Select {...args} concept={'normal'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>

    <Select {...args} concept={'transparent'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
  </div>
);

export const Modes: ComponentStory<typeof Select> = (args: any) => (
  <>
    <Select {...args} mode={'onwhite'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>

    <Select {...args} mode={'ongrey'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
    <Select {...args} mode={'onblueberry'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
    <Select {...args} mode={'oninvalid'} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
  </>
);

export const Width: ComponentStory<typeof Select> = (args: any) => (
  <>
    <Select {...args} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>

    <Select {...args} width={20} label={'Velg noe'} name={'select name 1'}>
      <option value={'Option 1'}>{'Option 1'}</option>
      <option value={'Option 2'}>{'Option 2'}</option>
      <option value={'Option 3'}>{'Option 3'}</option>
    </Select>
  </>
);

export const AllExamples: ComponentStory<typeof Select> = (args: any) => (
  <>
    <Default {...args} />
    <Concepts {...args} />
    <Modes {...args} />
    <Width {...args} />
  </>
);
