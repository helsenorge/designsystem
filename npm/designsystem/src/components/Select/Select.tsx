import React from 'react';

import classNames from 'classnames';

import selectStyles from './styles.module.scss';
import { AnalyticsId, AVERAGE_CHARACTER_WIDTH_PX, FormMode, IconSize } from '../../constants';
import Icon from '../Icons';
import ChevronDown from '../Icons/ChevronDown';
import { getColor } from '../../theme/currys';
import ErrorWrapper from '../ErrorWrapper';
import { useUuid } from '../../hooks/useUuid';

type SelectConcept = 'normal' | 'transparent';

export interface SelectProps extends Pick<React.SelectHTMLAttributes<HTMLSelectElement>, 'name' | 'disabled' | 'required' | 'value'> {
  /** Component shown after label */
  afterLabelChildren?: React.ReactNode;
  /** Sets the content of the select element. */
  children: React.ReactNode;
  /** Adds custom classes to the element. */
  className?: string;
  /** Changes the visuals of the component */
  concept?: SelectConcept;
  /** The label text above the select */
  label?: string;
  /** Changes the visuals of the component */
  mode?: keyof typeof FormMode;
  /** Activates Error style for the select component - This is can be true while errorText is empty, when in a FormGroup */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Sets the data-testid attribute. */
  testId?: string;
  /** select id of the select element */
  selectId?: string;
  /** Width of select in characters (approximate) */
  width?: number;
}

const getSelectMaxWidth = (characters: number): string => {
  const paddingWidth = '2rem';

  return `calc(${characters * AVERAGE_CHARACTER_WIDTH_PX}px + ${paddingWidth})`;
};

const getIconColor = (invalid: boolean, disabled: boolean) => {
  const iconColor = invalid ? 'cherry' : 'blueberry';
  return disabled ? getColor('neutral', 500) : getColor(iconColor, 600);
};

export const Select = React.forwardRef((props: SelectProps, ref: React.Ref<HTMLSelectElement>) => {
  const {
    afterLabelChildren,
    className,
    children,
    concept = 'normal',
    disabled,
    error,
    errorText,
    label,
    selectId,
    name = selectId,
    mode,
    testId,
    width,
    required,
    value,
  } = props;

  const uuid = useUuid(selectId);
  const onBlueberry = mode === 'onblueberry';
  const invalid = mode === 'oninvalid' || !!errorText || !!error;
  const onDark = mode === 'ondark';
  const iconColor = getIconColor(invalid, !!disabled);
  const maxWidth = width ? getSelectMaxWidth(width) : undefined;

  const selectLabelClasses = classNames(selectStyles['select-wrapper__label-wrapper'], {
    [selectStyles['select-wrapper__label-wrapper--on-dark']]: onDark,
  });

  const selectInnerWrapperClasses = classNames(
    selectStyles['select-inner-wrapper'],
    {
      [selectStyles['select-inner-wrapper--transparent']]: concept === 'transparent',
      [selectStyles['select-inner-wrapper--on-blueberry']]: onBlueberry,
      [selectStyles['select-inner-wrapper--invalid']]: invalid,
      [selectStyles['select-inner-wrapper--disabled']]: disabled,
    },
    className
  );

  const selectClasses = classNames(selectStyles.select, {
    [selectStyles['select--on-blueberry']]: onBlueberry,
    [selectStyles['select--invalid']]: invalid,
  });

  return (
    <ErrorWrapper errorText={errorText}>
      <div data-testid={testId} data-analyticsid={AnalyticsId.Select} className={selectStyles['select-wrapper']} style={{ maxWidth }}>
        {label && (
          <div className={selectLabelClasses}>
            <label htmlFor={uuid}>{label}</label>
            {afterLabelChildren && <div className={selectStyles['select-wrapper__after-label-children']}>{afterLabelChildren}</div>}
          </div>
        )}
        <div className={selectInnerWrapperClasses}>
          <Icon className={selectStyles['select-arrow']} svgIcon={ChevronDown} color={iconColor} size={IconSize.XSmall} />
          <select
            aria-invalid={!!invalid}
            id={uuid}
            name={name}
            className={selectClasses}
            disabled={disabled}
            ref={ref}
            required={required}
            value={value}
          >
            {children}
          </select>
        </div>
      </div>
    </ErrorWrapper>
  );
});

export default Select;
