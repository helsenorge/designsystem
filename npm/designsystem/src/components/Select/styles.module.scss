@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';
@import '../../scss/font-settings.scss';

.select-wrapper {
  display: flex;
  flex-direction: column;

  &__label-wrapper {
    display: inline-flex;
    align-items: center;
    line-height: getSpacer(m);
    font-size: $font-size-sm;
    font-weight: 600;
    padding-bottom: getSpacer(xs);

    @media (min-width: map-get($grid-breakpoints, md)) {
      line-height: getSpacer(l);
      font-size: $font-size-md;
    }

    &--on-dark {
      color: $white;
    }
  }

  &__after-label-children {
    padding-left: 0.75rem;
  }
}

.select-inner-wrapper {
  // Grid brukes til Ã¥ kunne plassere custom ikon fritt over select elementet
  display: inline-grid;
  background-color: $white;
  align-items: center;
  // Andre kolonne her styrer plassering av select ikonet
  grid-template-columns: 1fr 2.8rem;

  &--transparent {
    background-color: transparent;
  }

  &:hover {
    background: $neutral50;
  }
  &--on-blueberry {
    &:hover {
      background-color: $blueberry50;
    }
  }
  &--invalid {
    &:hover {
      background-color: $cherry50;
    }
  }
  &--disabled {
    background: $neutral50;

    &:hover {
      background: $neutral50;
    }
  }
}

.select {
  /* Pil fjernes for browsere utenom IE */
  appearance: none;

  height: 2.5rem;
  align-items: center;
  background-color: transparent;
  grid-column: 1 / 3;
  grid-row: 1;
  padding-left: getSpacer(xs);
  border: none;
  outline: $neutral700 solid 2px;
  outline-offset: -2px;
  font-family: inherit;
  font-size: $font-size-sm;
  line-height: $lineheight-size-sm;
  border-radius: 0;

  @media (min-width: map-get($grid-breakpoints, md)) {
    height: 3.125rem;
    font-size: $font-size-md;
  }

  &--on-blueberry {
    outline-color: $blueberry500;
  }
  &--invalid {
    outline-color: $cherry500;
  }

  :hover > & {
    outline-width: 3px;
    outline-offset: -1px;
  }
  &:focus {
    outline: $black solid 4px;
    outline-offset: -2px;
  }

  &:disabled {
    outline: $neutral500 solid 2px;
    outline-offset: -2px;
  }

  & > option {
    background: $white;
    font-family: inherit;
    font-size: $font-size-sm;

    @media (min-width: map-get($grid-breakpoints, md)) {
      font-size: $font-size-md;
    }
  }
}

.select-arrow {
  grid-column: 2;
  grid-row: 1;
}
