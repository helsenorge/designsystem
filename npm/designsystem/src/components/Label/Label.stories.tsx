import React from 'react';

import { StoryObj, Meta } from '@storybook/react';

import Label, { LabelText } from './Label';
import { Sublabel } from './SubLabel';
import { IconSize } from '../../constants';
import { getColor } from '../../theme/currys';
import Checkbox from '../Checkbox';
import Icon from '../Icon';
import HelpSign from '../Icons/HelpSign';
import Input from '../Input';
import RadioButton from '../RadioButton/RadioButton';
import Select from '../Select';
import StatusDot from '../StatusDot';
import Textarea from '../Textarea';

type LabelWithAndCustomArgs = React.ComponentProps<typeof Label> & {
  showIcon: boolean;
};

const meta = {
  title: '@helsenorge∕designsystem-react/Components/Label',
  component: Label,
  parameters: {
    docs: {
      description: {
        component:
          'Som en innbygger vil jeg kunne se flere dataelementer gruppert i en og samme label på en konsistent måte slik at jeg mer effektivt kan scanne og forstå valgene jeg skal kunne gjøre. \n\n <b>For å sette opp aria-label riktig med Label komponentet:</b> \n\n Gi input feltet sin id til id propen til Label. Gi Input feltet sin aria-describedby prop eventuelle sublabel/statusdot id(er). Du kan skjule bestemte label/sublabel tekster for skjermlesere ved hjelp av hideFromScreenReader property per tekst.',
      },
    },
  },
  argTypes: {},
} satisfies Meta<LabelWithAndCustomArgs>;

export default meta;

type Story = StoryObj<LabelWithAndCustomArgs>;

export const Default: Story = {
  render: args => {
    const sublabelId1 = 'sublabel-testid1';
    const statusDotId1 = 'statusdot-testid1';
    const sublabelId2 = 'sublabel-testid2';
    const statusDotId2 = 'statusdot-testid2';
    const sublabelId3 = 'sublabel-testid3';
    const statusDotId3 = 'statusdot-testid3';

    const sublabelTexts: LabelText[] = [
      { text: 'This is a normal sublabel', type: 'normal' },
      { text: 'Semibold sublabel - hidden from screen reader', type: 'semibold', hideFromScreenReader: true },
    ];

    return (
      <>
        <Input
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
                { text: 'this is hidden from screen readers', hideFromScreenReader: true },
              ]}
              sublabel={<Sublabel id={sublabelId1} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId1} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId1 + ' ' + statusDotId1}
        />
        <Checkbox
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
              ]}
              sublabel={<Sublabel id={sublabelId2} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId2} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId2 + ' ' + statusDotId2}
        />
        <RadioButton
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label' },
                { text: 'semibold returns', type: 'semibold' },
              ]}
              sublabel={<Sublabel id={sublabelId3} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId3} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          aria-describedby={sublabelId3 + ' ' + statusDotId3}
        />
      </>
    );
  },
};

export const OnDark: Story = {
  render: args => {
    const sublabelId = 'sublabel-testid';
    const sublabelTexts: LabelText[] = [
      { text: 'normal sublabel with a pretty long text', type: 'normal' },
      { text: 'Semibold sublabel', type: 'semibold' },
    ];
    const statusDotId = 'statusdot-testid';

    return (
      <div style={{ padding: '3rem', backgroundColor: getColor('blueberry', 600) }}>
        <Input
          label={
            <Label
              {...args}
              labelTexts={[
                { text: 'Semibold label', type: 'semibold' },
                { text: 'normal label', type: 'normal' },
                { text: 'semibold returns', type: 'semibold' },
                { text: 'another normal label', type: 'normal' },
              ]}
              sublabel={<Sublabel id={sublabelId} sublabelTexts={sublabelTexts} />}
              statusDot={<StatusDot id={statusDotId} text={'Statusdot text'} variant={'alert'} />}
            />
          }
          mode={'ondark'}
          aria-describedby={sublabelId + ' ' + statusDotId}
        />
      </div>
    );
  },
};

export const ChildrenAfterLabel: Story = {
  render: args => (
    <Input
      {...args}
      label={
        <Label
          labelTexts={[{ text: 'Skriv inn din tekst', type: 'semibold' }]}
          afterLabelChildren={<Icon size={IconSize.XSmall} svgIcon={HelpSign} />}
        />
      }
    />
  ),
};

export const LabelAsString: Story = {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  render: ({ showIcon, ...rest }) => (
    <>
      <Input {...rest} label={'Test label'} />
      <Textarea {...rest} label={'Test label'} />
      <Select {...rest} label={'Test label'}>
        <option value={'Option 1'}>{'Option 1'}</option>
        <option value={'Option 2'}>{'Option 2'}</option>
        <option value={'Option 3'}>{'Option 3'}</option>
      </Select>
      <Checkbox {...rest} label={'Test label'} />
      <RadioButton {...rest} label={'Test label'} />
    </>
  ),
};
