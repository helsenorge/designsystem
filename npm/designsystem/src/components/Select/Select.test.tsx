import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Select from './Select';
import { FormMode } from '../../constants';
import Label from '../Label';

describe('Gitt at Select skal vises', (): void => {
  describe('Når Select rendres', (): void => {
    test('Så vises Select', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const label = screen.getByText('Label test').parentElement;
      expect(label).toBeVisible();

      const select = screen.getByRole('combobox');
      expect(select).toBeVisible();
      expect(select).toHaveClass('select');

      const icon = select.previousElementSibling;
      expect(icon).toBeVisible();
      expect(icon).toHaveClass('hnds-style-icon select-arrow');
    });
  });

  describe('Når disabled er true', (): void => {
    test('Så vises Select som disabled', (): void => {
      render(
        <Select disabled selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const selectGrid = screen.getByRole('combobox').parentElement;
      expect(selectGrid).toHaveClass('select-inner-wrapper select-inner-wrapper--disabled');
    });
  });

  describe('Når mode er onBlueberry', (): void => {
    test('Så vises Select med onBlueberry styling', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} mode={FormMode.onblueberry}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const select = screen.getByRole('combobox');
      const selectGrid = select?.parentElement;

      expect(select).toHaveClass('select select--on-blueberry');
      expect(selectGrid).toHaveClass('select-inner-wrapper select-inner-wrapper--on-blueberry');
    });
  });

  describe('Når Select får satt error', (): void => {
    test('Så vises Select med indre error styling, uten ytre error styling', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} error>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const select = screen.getByRole('combobox');
      const wrapper = select?.parentElement;

      expect(select).toHaveClass('select select--invalid');
      expect(wrapper).toHaveClass('select-inner-wrapper select-inner-wrapper--invalid');
    });
  });

  describe('Når select får satt errorText', (): void => {
    test('Så vises Select med errormelding i tilleg til error styling', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} errorText={'error error!'}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      expect(screen.getByText('error error!')).toBeVisible();

      const select = screen.getByRole('combobox');
      const wrapper = select?.parentElement;

      expect(select).toHaveClass('select select--invalid');
      expect(wrapper).toHaveClass('select-inner-wrapper select-inner-wrapper--invalid');
    });
  });

  describe('Når name-prop er satt', (): void => {
    test('Så har Select riktig name', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} name={'custom-name'}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const select = screen.getByRole('combobox');
      expect(select).toHaveAttribute('name', 'custom-name');
    });
  });

  describe('Når required er satt', (): void => {
    test('Så er Select required', (): void => {
      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} required>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      const select = screen.getByRole('combobox');
      expect(select).toHaveAttribute('aria-required', 'true');
      expect(select).toBeRequired();
    });
  });

  describe('Når value og onChange er satt', (): void => {
    test('Så har select riktig value, og onChange kalles når man endrer', async (): Promise<void> => {
      const handleChange = jest.fn();

      render(
        <Select selectId={'test01'} label={<Label labelTexts={[{ text: 'Label test' }]} />} value={'Option 2'} onChange={handleChange}>
          <option value={'Option 1'}>{'Option 1'}</option>
          <option value={'Option 2'}>{'Option 2'}</option>
          <option value={'Option 3'}>{'Option 3'}</option>
        </Select>
      );

      expect(handleChange).not.toHaveBeenCalled();

      const select = screen.getByRole('combobox');
      expect(select).toHaveValue('Option 2');

      await userEvent.selectOptions(select, 'Option 1');

      expect(handleChange).toHaveBeenCalledTimes(1);

      await userEvent.selectOptions(select, 'Option 3');

      expect(handleChange).toHaveBeenCalledTimes(2);
    });
  });
});
