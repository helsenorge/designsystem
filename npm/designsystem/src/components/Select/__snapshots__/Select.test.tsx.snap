// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Select skal vises Når Select rendres Så vises Select 1`] = `
<div>
  <div
    class="error-wrapper"
  >
    <div
      class="select-wrapper"
      data-analyticsid="select"
    >
      <div
        class="select-wrapper__label-wrapper"
      >
        <label
          for="test01"
        >
          Label test
        </label>
      </div>
      <div
        class="select-inner-wrapper"
      >
        <svg
          aria-hidden="true"
          class="hnds-style-icon select-arrow"
          data-analyticsid="icon"
          fill="#126F87"
          focusable="false"
          height="38"
          role="presentation"
          style="min-width: 38px; min-height: 38px;"
          viewBox="0 0 48 48"
          width="38"
        >
          <path
            d="M32.577 17.885l1.793 1.779-10.371 10.451L13.63 19.664l1.793-1.779 8.576 8.644z"
          />
        </svg>
        <select
          aria-invalid="false"
          class="select"
          id="test01"
          name="test01"
        >
          <option
            value="Option 1"
          >
            Option 1
          </option>
          <option
            value="Option 2"
          >
            Option 2
          </option>
          <option
            value="Option 3"
          >
            Option 3
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;
